File: src/questions/parkinglot/java/entity/ParkingTicket.java
package questions.parkinglot.java.entity;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

public class ParkingTicket {
    private final String ticketId;
    private final Vehicle vehicle;
    private final LocalDateTime entryTime;
    private final ParkingSpot spot;
    private LocalDateTime exitTime;

    public ParkingTicket(String ticketId, Vehicle vehicle, ParkingSpot spot) {
        this.ticketId = ticketId;
        this.vehicle = vehicle;
        this.spot = spot;
        this.entryTime = LocalDateTime.now();
    }

    public String getTicketId() {
        return ticketId;
    }

    public ParkingSpot getSpot() {
        return spot;
    }

    public Vehicle getVehicle() {
        return vehicle;
    }

    public LocalDateTime getEntryTime() {
        return entryTime;
    }

    public LocalDateTime getExitTime() {
        return exitTime;
    }

    public void setExitTime(LocalDateTime exitTime) {
        this.exitTime = exitTime;
    }

    public long getTotalParkingDuration() {
        LocalDateTime exit = exitTime != null ? exitTime : LocalDateTime.now();
        long hours = ChronoUnit.HOURS.between(entryTime, exit);
        return Math.max(hours, 1); // Minimum charge for 1 hour
    }


}

File: src/questions/parkinglot/py/entity/ParkingTicket.py
from questions.py.entity.Vehicle import Vehicle
from datetime import datetime
from typing import Optional


class ParkingTicket:
    def __init__(self, ticket_id: str, vehicle: Vehicle, spot: 'ParkingSpot'):
        self._ticket_id = ticket_id
        self._vehicle = vehicle
        self._spot = spot
        self._entry_time = datetime.now()
        self._exit_time: Optional[datetime] = None

    def set_exit_time(self, exit_time: datetime) -> None:
        """Sets the exit time for the parking ticket."""
        self._exit_time = exit_time

    def get_duration_in_hours(self) -> float:
        """Calculates the duration of parking in hours."""
        exit = self._exit_time if self._exit_time else datetime.now()
        duration = exit - self._entry_time
        hours = int(duration.total_seconds() // 3600)
        return max(1, hours)  # Minimum charge for 1 hour

    @property
    def ticket_id(self) -> str:
        return self._ticket_id

    @property
    def vehicle(self) -> Vehicle:
        return self._vehicle

    @property
    def spot(self) -> 'ParkingSpot':
        return self._spot

    @property
    def entry_time(self) -> datetime:
        return self._entry_time

    @property
    def exit_time(self) -> Optional[datetime]:
        return self._exit_time
